### ISSUE TITLE
`.ToEPDString` produces ambiguous move representations

### ISSUE THREAD

**Issue by Anduin2017** [lynx-chess/Lynx#840]
I wrote the following code:

```csharp

public class ChessGame
{
    private readonly Engine _engine;

    public ChessGame()
    {
        var channel = Channel.CreateUnbounded<string>(new UnboundedChannelOptions
        {
            SingleReader = true,
            SingleWriter = true
        });
        _engine = new Engine(channel.Writer);
    }

    public string GetBestMove(string fen)
    {
        _engine.AdjustPosition($"position fen {fen}");
        return _engine.IDDFS(1, 10)
            .BestMove
            .ToEPDString();
    }
}

var fen = "2kr3r/p5p1/2p2P2/5n1p/1p6/1B6/PPP3PP/R1B1Q1K1 b - - 0 20";
var game = new ChessGame();
var best = game.GetBestMove(fen); // Re8
```

Then I started a game:

```
1. e4 e5 2. Nf3 d5 3. exd5 Qxd5 4. Nc3 Qe6 5. Bb5+ c6 6. Ba4 b5 7. Bb3 Qf6 8. O-O h5 9. d4 Nd7 10. dxe5 Nxe5 11. Nxe5 Qxe5 12. Re1 Bd6 13. Rxe5+ Bxe5 14. Qe1 f6 15. f4 b4 16. Nd1 Bf5 17. fxe5 O-O-O 18. Ne3 Ne7 19. Nxf5 Nxf5 20. exf6
```

Game fen is:

```
2kr3r/p5p1/2p2P2/5n1p/1p6/1B6/PPP3PP/R1B1Q1K1 b - - 0 20
```

However, in this situation:

![Screenshot from 2024-07-02 12-37-55](https://github.com/lynx-chess/Lynx/assets/19531547/f0f609e8-9c57-4e02-b5da-b1b884a76e4b)

Engine is responding: `Re8`.

`Re8` here is a conflict move and can't be understood.

**Comment by eduherminio** [comment]
Hi @Anduin2017, thanks for reporting this issue (and for using Lynx!).

The problem is with `.ToEPDString()`, representation method, the moves themselves aren't ambiguous.

This bug has been around since the beginning, but I'll try to look into it soon since you're also suffering from it now.

Is using `.ToUCIString()` an option for you? Although I'm aware that would imply changing whatever is interpreting those moves on your end.

**Comment by Anduin2017** [comment]
![image](https://github.com/lynx-chess/Lynx/assets/19531547/6df3496b-c614-4a4e-ba42-d577ebcb8896)

Sorry but I tried searching `.ToUCIString()` but found no result. Is it an alternative?

Thanks for the reply!!!

**Comment by eduherminio** [comment]
> Sorry but I tried searching `.ToUCIString()` but found no result. Is it actually `ToMoveString`? However my game system is already coupled with `PGN` and need standard PGN from the engine. :(

Sorry, I was on the phone, I meant `.UCIString()`, but as you mentioned that won't solve your issue because of your dependency on PGN/EPD

**Comment by eduherminio** [comment]
Give me some time, but I'll provide a fix it and expose a method that produces valid EPD move representation.
You'll need to provide the original position and It will be much costly than the current one, but that cost should be negligent compared to search times.

However, bear that mind that if your project grows in the future and manages to integrate with other chess engines, it'll have to communicate with them via standard I/O (i.e. running them using some .NET CLI wrapper) and, what is relevant to this issue, using UCI protocol.

Also, in your last screenshot you had my repo cloned. You _should_ be able to use [LynxChess.Lynx](https://www.nuget.org/packages/LynxChess.Lynx) NuGet package directly instead, but you would the the first one (other than my own [Lynx_AzureBotApi](https://github.com/lynx-chess/Lynx_AzureBotApi) FHL project, and that was using an old version of the project) so I would expect some rough edges to be discovered.

**Comment by eduherminio** [comment]
Could you try the following branch: `bugfix/ToEPDString`? If everything works as expected, I can put a pre-release NuGet version out so that you're not blocked by this issue.

Be aware of the existence of `Game.ResetCurrentPositionToBeforeSearchState()`, since search cancellation can make the internal Game's position end up in an inconsistent state.

You could either do:

```csharp
    public string GetBestMove(string fen)
    {
        _engine.AdjustPosition($"position fen {fen}");

        var result = _engine.IDDFS(1, 10);
        _engine.Game.ResetCurrentPositionToBeforeSearchState();

        return result
            .BestMove
            .ToEPDString(_engine.Game.CurrentPosition);
    }
```

or in case you ever use `Engine.BestMove()` for built-in time management, since that currently messes with the Game's internal position, you gotta save the initial position yourself:

```csharp
    public string GetBestMove(string fen)
    {
        _engine.AdjustPosition($"position fen {fen}");
        var positionClone = new Position(_engine.Game.CurrentPosition);

        return _engine.BestMove(new("go btime 50000 wtime 50000 winc 3000 binc 2000")) // or new("go depth 10")
            .BestMove
            .ToEPDString(positionClone);
    }
```

**Comment by Anduin2017** [comment]
Hi, @eduherminio !

I tried the following code:

```csharp
using Lynx;
using Lynx.Cli;
using Lynx.UCI.Commands.Engine;
using Microsoft.Extensions.Configuration;
using NLog;
using NLog.Extensions.Logging;
using System.Threading.Channels;
using Lynx.Model;

public class ChessGame
{
    private readonly Engine _engine;

    public ChessGame()
    {
        var channel = Channel.CreateUnbounded<string>(new UnboundedChannelOptions
        {
            SingleReader = true,
            SingleWriter = true
        });
        _engine = new Engine(channel.Writer);
    }

    public string GetBestMove(string fen)
    {
        _engine.AdjustPosition($"position fen {fen}");
        var positionClone = new Position(_engine.Game.CurrentPosition);

        var result = _engine.IDDFS(1, 10);
        _engine.Game.ResetCurrentPositionToBeforeSearchState();

        return result
            .BestMove
            .ToEPDString(positionClone);
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var fen = "2kr3r/p5p1/2p2P2/5n1p/1p6/1B6/PPP3PP/R1B1Q1K1 b - - 0 20";
        var game = new ChessGame();
        var best = game.GetBestMove(fen); // Re8
        Console.WriteLine(best);
    }
}

```

Output is `Rhe8`! It's working!!

Thank you so much for helping!!!

**Comment by Anduin2017** [comment]
By the way, I tried the new code. Here's another question:

Under this game:

```
3rr1k1/5ppp/p1p5/1P6/8/Q2B4/PK6/3b4 b - - 0 33
```

The code above will generate response:

```
ccxd5
```

However, for some system it couldn't understand it. For example, lichess.org prefer:

```
cxd5
```

![image](https://github.com/lynx-chess/Lynx/assets/19531547/0b7353bd-38f2-45a0-bf44-460f3b39e00d)

**Comment by eduherminio** [comment]
Good catch, I pushed a fix for pawn captures and made en-passant and castling moves notation compliant.
Give it a try and I'll release a NuGet package in 1-2 days if everything looks ok.

**Comment by Anduin2017** [comment]
```bash
anduin@u-real:~/Desktop/Lynx/src/Lynx.Cli$ cat ./Program.cs 
using Lynx;
using Lynx.Cli;
using Lynx.UCI.Commands.Engine;
using Microsoft.Extensions.Configuration;
using NLog;
using NLog.Extensions.Logging;
using System.Threading.Channels;
using Lynx.Model;

public class ChessGame
{
    private readonly Engine _engine;

    public ChessGame()
    {
        var channel = Channel.CreateUnbounded<string>(new UnboundedChannelOptions
        {
            SingleReader = true,
            SingleWriter = true
        });
        _engine = new Engine(channel.Writer);
    }

    public string GetBestMove(string fen)
    {
        _engine.AdjustPosition($"position fen {fen}");
        var positionClone = new Position(_engine.Game.CurrentPosition);

        var result = _engine.IDDFS(1, 10);
        _engine.Game.ResetCurrentPositionToBeforeSearchState();

        return result
            .BestMove
            .ToEPDString(positionClone);
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var fen = "3rr1k1/5ppp/p1p5/1P6/8/Q2B4/PK6/3b4 b - - 0 33";
        var game = new ChessGame();
        var best = game.GetBestMove(fen); // Re8
        Console.WriteLine(best);
    }
anduin@u-real:~/Desktop/Lynx/src/Lynx.Cli$ dotnet run
cxb5
anduin@u-real:~/Desktop/Lynx/src/Lynx.Cli$ git branch
* bugfix/ToEPDString
  main
anduin@u-real:~/Desktop/Lynx/src/Lynx.Cli$ 
```

I tried your code, works like a charm!!! Thank you so much!!

**Comment by eduherminio** [comment]
@Anduin2017 Fix available in [v1.6.0-alpha.1](https://www.nuget.org/packages/LynxChess.Lynx/1.6.0-alpha.1) NuGet package.
Do reach out again if you face any issues 😃
